Module code is executed twice when collecting doctests

- [x] a detailed description of the bug or problem you are having
- [x] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

### Bug description and minimal example

https://github.com/flying-sheep/pytest-doctest-import-twice

### Versions

OS: Arch Linux

```nushell
pip list --format=json --exclude-editable | from json | to md --pretty
```

| name       | version |
| ---------- | ------- |
| iniconfig  | 2.0.0   |
| packaging  | 23.1    |
| pip        | 23.2.1  |
| pluggy     | 1.2.0   |
| pytest     | 7.4.0   |
| setuptools | 68.0.0  |
| wheel      | 0.41.0  |

yes, I'm not saying this is caused by `PYTHONPATH`. But your example needs the `PYTHONPATH` set to run, even when using `hatch run`, at least on my system:

```
$ hatch run pytest
============================= test session starts ==============================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0
rootdir: /home/florian/tmp/11306/pytest-doctest-import-twice
configfile: pyproject.toml
collected 0 items / 3 errors                                                   

==================================== ERRORS ====================================
___________ ERROR collecting pytest_doctest_import_twice/__init__.py ___________
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/runner.py:372: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/doctest.py:547: in collect
    module = import_path(
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/pathlib.py:538: in import_path
    spec.loader.exec_module(mod)  # type: ignore[union-attr]
<frozen importlib._bootstrap_external>:940: in exec_module
    ???
<frozen importlib._bootstrap>:241: in _call_with_frames_removed
    ???
pytest_doctest_import_twice/__init__.py:1: in <module>
    from .singleton import MyVindictiveSingleton
E   ModuleNotFoundError: No module named 'pytest_doctest_import_twice'
_____________________ ERROR collecting tests/test_basic.py _____________________
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/runner.py:372: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/doctest.py:547: in collect
    module = import_path(
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/pathlib.py:538: in import_path
    spec.loader.exec_module(mod)  # type: ignore[union-attr]
../../../.local/share/hatch/env/virtual/pytest-doctest-import-twice/fxat3dFr/pytest-doctest-import-twice/lib/python3.11/site-packages/_pytest/assertion/rewrite.py:178: in exec_module
    exec(co, module.__dict__)
tests/test_basic.py:1: in <module>
    from pytest_doctest_import_twice import instance
E   ImportError: cannot import name 'instance' from 'pytest_doctest_import_twice' (unknown location)
_____________________ ERROR collecting tests/test_basic.py _____________________
ImportError while importing test module '/home/florian/tmp/11306/pytest-doctest-import-twice/tests/test_basic.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_basic.py:1: in <module>
    from pytest_doctest_import_twice import instance
E   ImportError: cannot import name 'instance' from 'pytest_doctest_import_twice' (unknown location)
=========================== short test summary info ============================
ERROR pytest_doctest_import_twice/__init__.py - ModuleNotFoundError: No module named 'pytest_doctest_import_twice'
ERROR tests/test_basic.py - ImportError: cannot import name 'instance' from 'pytest_doctest_import_twic...
ERROR tests/test_basic.py
!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 3 errors in 0.10s ===============================
```

--- a/changelog/11306.bugfix.rst
+++ b/changelog/11306.bugfix.rst
@@ -1,0 +1,1 @@
+Fixed bug using ``--importmode=importlib`` which would cause package ``__init__.py`` files to be imported more than once in some cases.

--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -623,6 +623,10 @@ def module_name_from_path(path: Path, root: Path) -> str:
         # Use the parts for the relative path to the root path.
         path_parts = relative_path.parts
 
+    # Module name for packages do not contain the __init__ file.
+    if path_parts[-1] == "__init__":
+        path_parts = path_parts[:-1]
+
     return ".".join(path_parts)
 
 

--- a/testing/test_pathlib.py
+++ b/testing/test_pathlib.py
@@ -18,6 +18,7 @@ from _pytest.pathlib import fnmatch_ex
 from _pytest.pathlib import get_extended_length_path_str
 from _pytest.pathlib import get_lock_path
 from _pytest.pathlib import import_path
+from _pytest.pathlib import ImportMode
 from _pytest.pathlib import ImportPathMismatchError
 from _pytest.pathlib import insert_missing_modules
 from _pytest.pathlib import maybe_delete_a_numbered_dir
@@ -585,6 +586,10 @@ class TestImportLibMode:
         result = module_name_from_path(Path("/home/foo/test_foo.py"), Path("/bar"))
         assert result == "home.foo.test_foo"
 
+        # Importing __init__.py files should return the package as module name.
+        result = module_name_from_path(tmp_path / "src/app/__init__.py", tmp_path)
+        assert result == "src.app"
+
     def test_insert_missing_modules(
         self, monkeypatch: MonkeyPatch, tmp_path: Path
     ) -> None:
@@ -615,3 +620,43 @@ class TestImportLibMode:
         assert sorted(modules) == ["xxx", "xxx.tests", "xxx.tests.foo"]
         assert modules["xxx"].tests is modules["xxx.tests"]
         assert modules["xxx.tests"].foo is modules["xxx.tests.foo"]
+
+    def test_importlib_package(self, monkeypatch: MonkeyPatch, tmp_path: Path):
+        """
+        Importing a package using --importmode=importlib should not import the
+        package's __init__.py file more than once (#11306).
+        """
+        monkeypatch.chdir(tmp_path)
+        monkeypatch.syspath_prepend(tmp_path)
+
+        package_name = "importlib_import_package"
+        tmp_path.joinpath(package_name).mkdir()
+        init = tmp_path.joinpath(f"{package_name}/__init__.py")
+        init.write_text(
+            dedent(
+                """
+                from .singleton import Singleton
+
+                instance = Singleton()
+                """
+            ),
+            encoding="ascii",
+        )
+        singleton = tmp_path.joinpath(f"{package_name}/singleton.py")
+        singleton.write_text(
+            dedent(
+                """
+                class Singleton:
+                    INSTANCES = []
+
+                    def __init__(self) -> None:
+                        self.INSTANCES.append(self)
+                        if len(self.INSTANCES) > 1:
+                            raise RuntimeError("Already initialized")
+                """
+            ),
+            encoding="ascii",
+        )
+
+        mod = import_path(init, root=tmp_path, mode=ImportMode.importlib)
+        assert len(mod.instance.INSTANCES) == 1

